// guestbell-forms/build/components/money/Money.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Money = exports.MoneyRaw = void 0;

var React = _interopRequireWildcard(require("react"));

var _Select = require("../select/Select");

var _Text = require("../text/Text");

var _InputGroup = _interopRequireDefault(require("../inputGroup/InputGroup"));

var _BaseInput2 = require("../base/input/BaseInput");

var _Button = require("../button/Button");

var _withFormContext = require("../form/withFormContext");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var PlusIcon = function PlusIcon(props) {
  return React.createElement("svg", props, React.createElement("path", {
    d: "M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
  }));
};

PlusIcon.defaultProps = {
  xmlns: "http://www.w3.org/2000/svg",
  width: "24",
  height: "24",
  viewBox: "0 0 24 24"
}; // Libs

var MoneyRaw =
/*#__PURE__*/
function (_BaseInput) {
  _inherits(MoneyRaw, _BaseInput);

  function MoneyRaw(props) {
    var _this;

    _classCallCheck(this, MoneyRaw);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(MoneyRaw).call(this, props, false));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "onCurrencyChanged", function (index, currentCurrencies) {
      return function (e) {
        var newPrices = [].concat(_this.props.prices);
        newPrices[index].currency = currentCurrencies.filter(function (cc) {
          return cc.value.toString() === e.target.value;
        })[0];

        _this.props.onPricesChange(newPrices);
      };
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "onTheFlightValidate", function (value) {
      var num = Number(value);
      var parts = value.split('.');

      if (parts && parts.length > 1 && parts[parts.length - 1].length > 2) {
        return false;
      }

      if (!isNaN(num)) {
        return true;
      }

      if (num) {
        return false;
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "onPriceChanged", function (index) {
      return function (e) {
        var newPrices = [].concat(_this.props.prices);
        var str = e.target.value;
        var num = Number(str);

        if (!isNaN(num)) {
          newPrices[index].value = num;
        }

        _this.props.onPricesChange(newPrices);

        if (!_this.state.touched) {
          _this.touch();
        }
      };
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "removePriceClick", function (index) {
      return function () {
        var newPrices = _this.props.prices.filter(function (price, itemIndex) {
          return itemIndex !== index;
        });

        _this.props.onPricesChange(newPrices);

        if (newPrices.length === 0 && _this.props.required) {
          _this.setInvalid(['Required']);
        }

        if (!_this.state.touched) {
          _this.touch();
        }
      };
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "addPriceClick", function (unusedCurrencies) {
      return function () {
        _this.props.onPricesChange(_this.props.prices.concat([{
          value: 0,
          currency: unusedCurrencies[0]
        }]));

        _this.setValid();

        if (!_this.state.touched) {
          _this.touch();
        }
      };
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "onFocus", function () {
      return _this.setState({
        focused: true
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "onBlur", function () {
      return _this.setState({
        focused: false
      });
    });

    if (!props.currencies || !props.currencies.length) {
      throw Error('No currencies supplied to money input');
    }

    _this.state = Object.assign(_this.state, {
      isValid: props.required ? props.prices.length > 0 : true,
      errors: props.required && props.prices.length === 0 ? ['Required'] : [],
      handleValueChangeEnabled: false
    });

    _this.subscribeSelf(props);

    return _this;
  }

  _createClass(MoneyRaw, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var unusedCurrencies = this.props.currencies;
      return React.createElement(_InputGroup.default, {
        title: this.props.title
      }, React.createElement("div", {
        className: 'input__base money-input ' + this.getValidationClass() + (this.props.className ? ' ' + this.props.className : ''),
        ref: this.containerRef
      }, this.props.prices && this.props.prices.map(function (item, index) {
        var currentCurrencies = _this2.props.currencies.filter(function (currency) {
          return _this2.props.prices.filter(function (priceCurrency, priceIndex) {
            return priceIndex !== index && priceCurrency.currency.value === currency.value;
          }).length === 0;
        });

        var retComponents = currentCurrencies.length ? React.createElement("div", {
          key: index
        }, React.createElement(_Select.Select, {
          onFocus: _this2.onFocus,
          onBlur: _this2.onBlur,
          className: 'money-input__select m-0',
          values: currentCurrencies,
          onChange: _this2.onCurrencyChanged(index, currentCurrencies),
          value: item.currency.value.toString()
        }), React.createElement(_Text.Text, {
          onFocus: _this2.onFocus,
          onBlur: _this2.onBlur,
          onTheFlightValidate: _this2.onTheFlightValidate,
          placeholder: '0.00',
          className: 'money-input__text m-0',
          validators: ['number'],
          value: item.value ? item.value.toString() : '',
          onChange: _this2.onPriceChanged(index),
          type: "number"
        }), _this2.props.prices.length > 0 && React.createElement(_Button.Button, {
          blank: true,
          type: "error",
          onClick: _this2.removePriceClick(index),
          className: "transform-rotate--45 line-height--0",
          buttonProps: {
            title: 'Remove price'
          },
          circular: true
        }, React.createElement(PlusIcon, null))) : null;
        unusedCurrencies = unusedCurrencies.filter(function (currency) {
          return currency.value !== item.currency.value;
        });
        return retComponents;
      }), (this.props.allowMultiple || this.props.prices && !this.props.prices.length) && unusedCurrencies.length ? React.createElement(_Button.Button, {
        blank: true,
        type: "primary",
        className: "line-height--0",
        onClick: this.addPriceClick(unusedCurrencies),
        circular: true,
        buttonProps: {
          title: this.props.prices && this.props.prices.length === 0 ? 'Add price' : 'Add new currency'
        }
      }, React.createElement(PlusIcon, null)) : null, React.createElement("span", {
        className: "highlight"
      }), React.createElement("span", {
        className: 'bar ' + (this.state.focused ? 'focused' : '')
      }), this.renderDefaultValidation(), this.props.label && React.createElement("label", {
        className: this.props.prices && this.props.prices ? 'label--focused' : ''
      }, this.renderLabel())));
    }
  }]);

  return MoneyRaw;
}(_BaseInput2.BaseInput);

exports.MoneyRaw = MoneyRaw;

_defineProperty(MoneyRaw, "defaultProps", Object.assign({}, _BaseInput2.BaseInput.defaultProps, {
  type: 'money',
  allowMultiple: false
}));

var Money = (0, _withFormContext.withFormContext)(MoneyRaw);
exports.Money = Money;
var _default = Money;
exports.default = _default;
