// guestbell-forms/build/components/select/Select.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Select = void 0;

var React = _interopRequireWildcard(require("react"));

var _InputGroup = _interopRequireDefault(require("../inputGroup/InputGroup"));

var _BaseInput2 = require("../base/input/BaseInput");

var _Button = require("../button/Button");

var _withFormContext = require("../form/withFormContext");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var PlusIcon = function PlusIcon(props) {
  return React.createElement("svg", props, React.createElement("path", {
    d: "M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
  }));
};

PlusIcon.defaultProps = {
  xmlns: "http://www.w3.org/2000/svg",
  width: "24",
  height: "24",
  viewBox: "0 0 24 24"
};

var SelectRaw =
/*#__PURE__*/
function (_BaseInput) {
  _inherits(SelectRaw, _BaseInput);

  function SelectRaw(props) {
    var _this;

    _classCallCheck(this, SelectRaw);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(SelectRaw).call(this, props));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "removeItemClick", function (item) {
      return function () {
        var newValues = _this.props.selectedValues.filter(function (sv) {
          return sv.value !== item.value;
        });

        _this.handleValid(newValues);

        _this.props.onSelectedValuesChange && _this.props.onSelectedValuesChange(newValues);
      };
    });

    var val = !props.value ? props.defaultEmpty ? '' : props.values && props.values.length > 0 ? props.values[0] : '' : props.value;
    _this.state = Object.assign(_this.state, {
      value: val
    });
    _this.handleChangeCustom = _this.handleChangeCustom.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    return _this;
  }

  _createClass(SelectRaw, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.props.multiple && this.handleValid(this.props.selectedValues);
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var finalValues = this.props.multiple ? this.props.values.filter(function (item) {
        return _this2.props.selectedValues.findIndex(function (t) {
          return t.value === item.value;
        }) < 0;
      }) : this.props.values;
      return React.createElement(_InputGroup.default, {
        title: this.props.title
      }, React.createElement("div", {
        className: 'input__base select-input ' + this.getValidationClass() + ' ' + (this.props.className ? this.props.className : '') + ' ' + (this.props.readOnly ? 'readonly' : '') + ' ' + (this.props.multiple ? 'multiple' : ''),
        ref: this.containerRef
      }, this.renderSelectedValues(), finalValues.length > 0 && (this.props.multiple && !this.props.readOnly || !this.props.multiple) && React.createElement("div", {
        className: "select-input__select__wrapper"
      }, !this.props.multiple && !this.props.readOnly || this.props.multiple ? React.createElement("select", {
        ref: this.inputRef,
        disabled: this.getDisabled(),
        required: this.props.required,
        onChange: this.handleChangeCustom,
        value: this.state.value,
        className: 'select-input__select ' + (this.state.value !== '' || this.props.selectedValues && this.props.selectedValues.length > 0 ? 'filled' : ''),
        onBlur: this.handleBlur,
        onFocus: this.handleFocus
      }, this.props.defaultEmpty && React.createElement("option", {
        key: -1,
        disabled: true,
        value: "",
        style: {
          display: 'none'
        }
      }), finalValues.map(function (value, index) {
        return React.createElement("option", {
          key: index,
          value: value.value
        }, value.label ? value.label : value.value);
      })) : React.createElement("span", {
        className: 'select-input__select ' + (this.state.value !== '' || this.props.selectedValues && this.props.selectedValues.length > 0 ? 'filled' : '')
      }, this.renderReadonly()), React.createElement("span", {
        className: "highlight"
      }), React.createElement("span", {
        className: "bar"
      }), this.renderDefaultValidation(), this.props.label && finalValues.length > 0 && React.createElement("label", null, this.renderLabel()))));
    }
  }, {
    key: "handleChangeCustom",
    value: function handleChangeCustom(event) {
      if (this.props.multiple) {
        var _value = event.target.value;
        var val = this.props.values.filter(function (item) {
          return item.value === _value;
        })[0];

        if (!val) {
          if (!isNaN(Number(_value))) {
            var valNumber = Number(_value);
            val = this.props.values.filter(function (item) {
              return item.value === valNumber;
            })[0];
          }
        }

        var newValues = this.props.selectedValues.concat(val);

        if (val) {
          this.props.onSelectedValuesChange && this.props.onSelectedValuesChange(newValues);
          this.handleValid(newValues);
          this.setState({
            value: ''
          });
        }
      } else {
        this.handleChange(event);
      }
    }
  }, {
    key: "handleValid",
    value: function handleValid(newValues) {
      if (this.props.required) {
        if (newValues.length > 0) {
          this.setValid();
        } else {
          this.setInvalid(['Required']);
        }
      }
    }
  }, {
    key: "renderReadonly",
    value: function renderReadonly() {
      var _this3 = this;

      var value = this.props.values.filter(function (item) {
        return item.value.toString() === _this3.state.value;
      })[0];
      return value ? value.label ? value.label : value.value : '';
    }
  }, {
    key: "renderSelectedValues",
    value: function renderSelectedValues() {
      var _this4 = this;

      return this.props.multiple ? this.props.selectedValues.length > 0 ? React.createElement("div", {
        className: "select-input__selectedValue__wrapper"
      }, this.props.selectedValues.map(function (item, index) {
        return React.createElement("div", {
          className: "select-input__selectedValue",
          key: index
        }, item.label ? item.label : item.value, !_this4.props.readOnly && React.createElement(_Button.Button, {
          disabled: item.forceSelected,
          circular: true,
          blank: true,
          onClick: _this4.removeItemClick(item),
          className: "ml-1 transform-rotate--45 line-height--0 p-0"
        }, React.createElement(PlusIcon, null)));
      })) : this.props.readOnly && React.createElement("div", {
        className: "select-input__selectedValue__wrapper"
      }, React.createElement("div", {
        className: "select-input__selectedValue"
      }, this.props.readonlyEmptyPlaceholder)) : null;
    }
  }]);

  return SelectRaw;
}(_BaseInput2.BaseInput);

_defineProperty(SelectRaw, "defaultProps", Object.assign({}, _BaseInput2.BaseInput.defaultProps, {
  defaultEmpty: true,
  multiple: false,
  readOnly: false,
  readonlyEmptyPlaceholder: 'N/A'
}));

var Select = (0, _withFormContext.withFormContext)(SelectRaw);
exports.Select = Select;
var _default = Select;
exports.default = _default;
