// guestbell-forms/build/components/form/FormValidationSummary.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FormValidationSummary = exports.FormValidationSummaryRaw = exports.DefaultComponent = void 0;

var React = _interopRequireWildcard(require("react"));

var _Guid = _interopRequireDefault(require("../utils/Guid"));

var _withFormContext = require("./withFormContext");

var _Button = require("./../button/Button");

var _FormValidationSummary = require("./FormValidationSummary");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var classNames = require('classnames');

var ArrowIcon = function ArrowIcon(props) {
  return React.createElement("svg", props, React.createElement("path", {
    d: "M12 8c-2.21 0-4 1.79-4 4s1.79 4 4 4 4-1.79 4-4-1.79-4-4-4zm8.94 3A8.994 8.994 0 0 0 13 3.06V1h-2v2.06A8.994 8.994 0 0 0 3.06 11H1v2h2.06A8.994 8.994 0 0 0 11 20.94V23h2v-2.06A8.994 8.994 0 0 0 20.94 13H23v-2h-2.06zM12 19c-3.87 0-7-3.13-7-7s3.13-7 7-7 7 3.13 7 7-3.13 7-7 7z"
  }));
};

ArrowIcon.defaultProps = {
  xmlns: "http://www.w3.org/2000/svg",
  width: "24",
  height: "24",
  viewBox: "0 0 24 24"
};

var DefaultComponent = function DefaultComponent(props) {
  return React.createElement(React.Fragment, null, props.componentsWithErrors.map(function (component, index) {
    return React.createElement("div", {
      key: index,
      className: "formValidationSummary__item"
    }, React.createElement("div", {
      className: "formValidationSummary__item__header"
    }, component.validation.name ? component.validation.name : 'unknown'), React.createElement("div", {
      className: "formValidationSummary__item__errors"
    }, component.validation.errors && component.validation.errors.map(function (err, errIndex) {
      return React.createElement("span", {
        className: "formValidationSummary__item__error",
        key: errIndex
      }, err);
    })), React.createElement("div", {
      style: {
        flex: 'auto'
      }
    }), component.componentApi && component.componentApi.focus && component.componentApi.scrollTo && React.createElement(_Button.Button, {
      icon: true,
      className: "formValidationSummary__item__locate-button",
      circular: true,
      type: "error",
      small: true,
      blank: true,
      onClick: function onClick() {
        component.componentApi.scrollTo();
        component.componentApi.focus();
      }
    }, React.createElement(ArrowIcon, null)));
  }));
};

exports.DefaultComponent = DefaultComponent;

var FormValidationSummaryRaw =
/*#__PURE__*/
function (_React$PureComponent) {
  _inherits(FormValidationSummaryRaw, _React$PureComponent);

  function FormValidationSummaryRaw() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, FormValidationSummaryRaw);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(FormValidationSummaryRaw)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "componentId", (0, _Guid.default)());

    return _this;
  }

  _createClass(FormValidationSummaryRaw, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var componentsWithErrors = this.props.formContext ? Object.keys(this.props.formContext.components).map(function (key) {
        return _this2.props.formContext.components[key];
      }).filter(function (component) {
        if (!component.validation.isValid && !component.validation.name) {
          console.warn(component, 'Has validation error. FormValidationSummary is present but the component has no validation name');
        }

        return !component.validation.isValid;
      }) : [];
      var containerClassName = classNames(this.props.containerClassName);
      var headerClassName = classNames(this.props.headerClassName);
      var footerClassName = classNames(this.props.footerClassName);
      return componentsWithErrors.length > 0 && React.createElement("div", {
        className: containerClassName
      }, this.props.title && React.createElement("div", {
        className: headerClassName
      }, this.props.title), React.createElement(this.props.Component, {
        componentsWithErrors: componentsWithErrors
      }), this.props.footer && React.createElement("div", {
        className: footerClassName
      }, this.props.footer));
    }
  }]);

  return FormValidationSummaryRaw;
}(React.PureComponent);

exports.FormValidationSummaryRaw = FormValidationSummaryRaw;

_defineProperty(FormValidationSummaryRaw, "defaultProps", {
  title: 'Hang on',
  footer: '... needs fixing',
  Component: DefaultComponent,
  containerClassName: 'formValidationSummary',
  headerClassName: 'formValidationSummary__header',
  footerClassName: 'formValidationSummary__footer'
});

var FormValidationSummary = (0, _withFormContext.withFormContext)(FormValidationSummaryRaw);
exports.FormValidationSummary = FormValidationSummary;
