// guestbell-forms/build/components/openingHours/openingHoursSpecial/OpeningHoursSpecial.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.OpeningHoursSpecial = exports.OpeningHoursSpecialRaw = exports.DateInput = void 0;

var React = _interopRequireWildcard(require("react"));

var _OpeningHoursUtil = _interopRequireDefault(require("../utils/OpeningHoursUtil"));

var _OpeningHoursDay = require("../openingHoursDay/OpeningHoursDay");

var _BaseInput2 = require("../../base/input/BaseInput");

var _Button = require("../../button/Button");

var _withFormContext = require("../../form/withFormContext");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var PlusIcon = function PlusIcon(props) {
  return React.createElement("svg", props, React.createElement("path", {
    d: "M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
  }));
};

PlusIcon.defaultProps = {
  xmlns: "http://www.w3.org/2000/svg",
  width: "24",
  height: "24",
  viewBox: "0 0 24 24"
};

var DateIcon = function DateIcon(props) {
  return React.createElement("svg", props, React.createElement("path", {
    d: "M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19a2 2 0 0 0 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
  }));
};

DateIcon.defaultProps = {
  xmlns: "http://www.w3.org/2000/svg",
  width: "24",
  height: "24",
  viewBox: "0 0 24 24"
}; // Libs

try {
  var Moment = require('moment');
} catch (_unused) {
  Moment = undefined;
}

try {
  var DatePicker = require('react-datepicker').default;
} catch (_unused2) {
  DatePicker = undefined;
} // Misc


var DAY_FORMAT = 'D/M/YYYY';

var DateInput =
/*#__PURE__*/
function (_React$PureComponent) {
  _inherits(DateInput, _React$PureComponent);

  function DateInput() {
    _classCallCheck(this, DateInput);

    return _possibleConstructorReturn(this, _getPrototypeOf(DateInput).apply(this, arguments));
  }

  _createClass(DateInput, [{
    key: "render",
    value: function render() {
      return React.createElement(_Button.Button, {
        blank: true,
        type: "primary",
        onClick: this.props.onClick
      }, React.createElement("div", {
        style: {
          display: 'flex'
        }
      }, React.createElement(DateIcon, {
        style: {
          marginRight: 4
        }
      }), this.props.children ? this.props.children : this.props.value));
    }
  }]);

  return DateInput;
}(React.PureComponent);

exports.DateInput = DateInput;

var OpeningHoursSpecialRaw =
/*#__PURE__*/
function (_BaseInput) {
  _inherits(OpeningHoursSpecialRaw, _BaseInput);

  function OpeningHoursSpecialRaw(props) {
    var _this;

    _classCallCheck(this, OpeningHoursSpecialRaw);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(OpeningHoursSpecialRaw).call(this, props));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "removeDayClick", function (index) {
      return function () {
        return _this.props.onDaysChange(_this.props.days.filter(function (d, indexInner) {
          return indexInner !== index;
        }));
      };
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "dateChanged", function (index, day) {
      return function (date) {
        var days = _this.props.days.slice(0);

        days[index] = _objectSpread({}, day, {
          date: date.toDate()
        });

        _this.props.onDaysChange(days);
      };
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "openingHoursChanged", function (index, day) {
      return function (openingHours) {
        var days = _this.props.days.slice(0);

        days[index] = _objectSpread({}, day, openingHours);

        _this.props.onDaysChange(days);
      };
    });

    return _this;
  }

  _createClass(OpeningHoursSpecialRaw, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      if (!DatePicker) {
        throw new Error('You need to install react-datepicker in order to use special day picker');
      }

      if (!Moment) {
        throw new Error('You need to install moment in order to use special day picker');
      }

      return React.createElement("div", {
        className: 'input__base openingHoursSpecial-input ' + this.getValidationClass() + ' ' + (this.props.className ? this.props.className : ''),
        ref: this.containerRef
      }, this.props.days.map(function (day, index) {
        return React.createElement(_OpeningHoursDay.OpeningHoursDay, {
          key: index,
          label: React.createElement("span", null, _OpeningHoursUtil.default.getLabelSuffix(day), React.createElement("span", {
            className: "float-right"
          }, React.createElement(_Button.Button, {
            blank: true,
            className: "openingHoursSpecial-input__button openingHoursSpecial-input__button--remove",
            onClick: _this2.removeDayClick(index)
          }, React.createElement(PlusIcon, null)))),
          openingHours: {
            times: day.times
          },
          onOpeningHoursChange: _this2.openingHoursChanged(index, day),
          title: React.createElement(DatePicker, {
            customInput: React.createElement(DateInput, null, !day.date && 'Choose date'),
            placeholder: _this2.props.placeholder,
            selected: day.date && Moment(day.date),
            dateFormat: DAY_FORMAT,
            onChange: _this2.dateChanged(index, day),
            excludeDates: _this2.props.days.filter(function (d) {
              return d.date;
            }).map(function (d) {
              return Moment(d.date);
            }),
            withPortal: true,
            minDate: Moment()
          })
        });
      }), React.createElement("span", {
        className: "bar"
      }), this.renderDefaultValidation());
    }
  }]);

  return OpeningHoursSpecialRaw;
}(_BaseInput2.BaseInput);

exports.OpeningHoursSpecialRaw = OpeningHoursSpecialRaw;

_defineProperty(OpeningHoursSpecialRaw, "defaultProps", Object.assign({}, _BaseInput2.BaseInput.defaultProps, {
  type: 'openingHoursSpecial',
  placeholder: DAY_FORMAT
}));

var OpeningHoursSpecial = (0, _withFormContext.withFormContext)(OpeningHoursSpecialRaw);
exports.OpeningHoursSpecial = OpeningHoursSpecial;
var _default = OpeningHoursSpecial;
exports.default = _default;
